### DISCUSSION TITLE
Error: Could not find or load main class bms.player.beatoraja.MainLoader

### DISCUSSION THREAD

**Discussion by kxrider** [wcko87/beatoraja-english-guide#35] [beatoraja questions / troubleshooting]
Carrying over from https://github.com/exch-bms2/beatoraja/issues/627.

**Comment by wcko87** [comment]
So from what I know, as long as you're getting the "Could not find or load main class bms.player.beatoraja.MainLoader" error, you have a missing javafx.
My first suspicion is that you might have multiple installations of Java installed in your PC, and the wrong installation of java is running by default.

**Reply by kxrider** [reply]
I certainly do have multiple installations of Java. How can I ensure the correct one is running?

**Reply by wcko87** [reply]
could you open a command window (Win+R, type cmd, enter), and type java -version?

**Reply by kxrider** [reply]
openjdk version "1.8.0_292"
OpenJDK Runtime Environment (build 1.8.0_292-b10)
OpenJDK 64-Bit Server VM (build 25.292-b10, mixed mode)

**Reply by wcko87** [reply]
alright, that is an installation of Java (OpenJDK) that doesn't have JavaFX.
The build from java.com and liberica's java should have JavaFX, while OpenJDK doesn't.
One way is to uninstall your OpenJDK java.

If you prefer not to uninstall that, the other option would be to manually change your environment variables to point to the correct installation of java (search "environment variables" in the start menu)

**Reply by kxrider** [reply]
I'd prefer the latter solution. I have a JAVA_HOME variable which already points to my Liberica JDK installation, so I am not sure why I get OpenJDK when doing java -version.

**Reply by wcko87** [reply]
it's not the JAVA_HOME variable that matters, it's the java that's in your PATH variable
that determines which installation of java runs in the command line by default
guess I didn't specify that clearly.

**Reply by kxrider** [reply]
I see. So, there are several javas in my path variable. LibericaJDK-8\bin\ is at the top.

**Reply by wcko87** [reply]
yeah, try cleaning it up, by deleting all the javas in the path variable except the one that you want to run by default

**Reply by wcko87** [reply]
you'll need to close and re-open your command prompt for your changes to take effect

**Reply by kxrider** [reply]
I removed any java related paths other than Liberica JDK 8 and restarted my computer. I still see OpenJDK 8 when I type java -version and get the javafx error when I run beatoraja.

**Reply by wcko87** [reply]
There are two path variables in the environment variables menu, one under "admin" and one under "system variables", have you checked both of them?

**Reply by kxrider** [reply]
Yes, most of the java paths were under "system variables" but I just now removed one from "user variables" for Intellij for good measure. Same result when running java -version and beatoraja.

**Reply by wcko87** [reply]
I'm pretty sure you've missed one of them, because there's no way java would run from the command line if it wasn't in either of the PATH variables.
I'd ask to have a look at a screenshot to see if I can spot it, though I'm not sure if posting a screenshot of that in a public place is the best idea.

**Reply by wcko87** [reply]
Oh, there is another, more direct alternative if you want to manage multiple installations of java.
Edit your beatoraja-config.bat file, and instead of `java -Xms1g -Xmx4g etc etc etc`, replace `java` with the direct path to the liberica java installation (like `C:\Program Files\Java\jdk1.8.0_231\bin\java.exe -Xms1g -Xmx4g etc etc etc`)

This way you can use the liberica java only for beatoraja, and use what you normally use for everything else.

**Reply by kxrider** [reply]
Okay, so I just removed Liberica JDK-8 from my PATH, and sure enough, the java command is no longer recognized on cmd, so I guess it was showing up as OpenJDK? But Liberica comes with JavaFX, so I'm not sure what's going on anymore.

**Reply by kxrider** [reply]
> Oh, there is another, more direct alternative if you want to manage multiple installations of java.
> Edit your beatoraja-config.bat file, and instead of `java -Xms1g -Xmx4g etc etc etc`, replace `java` with the direct path to the liberica java installation (like `C:\Program Files\Java\jdk1.8.0_231\bin\java.exe -Xms1g -Xmx4g etc etc etc`)
> 
> This way you can use the liberica java only for beatoraja, and use what you normally use for everything else.

Just tried this. Same error when replacing java with C:\"Program Files"\BellSoft\LibericaJDK-8\bin

**Reply by wcko87** [reply]
hmmm, that's interesting.
Liberica should indeed come with javafx, but I'm not 100% sure. Let me check.

**Reply by wcko87** [reply]
could I get a full error message for the Mainloader error?

**Reply by kxrider** [reply]
Sure:
```
Picked up _JAVA_OPTIONS: '-Dsun.java2d.opengl=true -Dawt.useSystemAAFontSettings=on -Dswing.aatext=true -Dswing.defaultlaf=com.sun.java.swing.plaf.gtk.GTKLookAndFeel' -Dfile.encoding="UTF-8"
Error: Could not find or load main class Files\BellSoft\LibericaJDK-8\bin\java.exe  -Xms1g -Xmx4g -cp beatoraja.jar;ir.* bms.player.beatoraja.MainLoader
```
And an update, I just reinstalled regular java 8 (which is supposed to come with javafx), and I guess after cleaning out my path and everything, that seems to have solved my problem, so thank you!

**Reply by wcko87** [reply]
glad to know that reinstalling the regular java 8 fixed your issue.
I'll have to look into liberica again, I've heard that it does have javafx by default, but I've never actually tried installing it.

**Reply by mak1non** [reply]
Liberica JDK has multiple package, and JavaFX exists only in "Full" package. Probably you used "Standard" or "Lite" one.
If you want to run beatoraja in Liberica JDK, you should use "Full JDK" or "Full JRE" package.

**Reply by wcko87** [reply]
@mak1non thanks a lot for that information! That indeed seems to be the case.
https://bell-sw.com/announcements/2020/01/15/Liberica-JDK-8u242-11.0.6-13.0.2/
I'll update the guide accordingly.
